src/index.rs

```diff
@@ -68,95 +93,42 @@ impl ResourceIndex {
         index
     }
 
+    /// Loads a previously stored resource index from the root path
+    ///
+    /// This function reads the index from the file system and returns a new
+    /// [`ResourceIndex`] instance. It looks for the index fie in
+    /// `root_path/.ark/index`
     pub fn load<P: AsRef<Path>>(root_path: P) -> Result<Self> {
-        let root_path: PathBuf = root_path.as_ref().to_owned();
-
-        let index_path: PathBuf = root_path.join(ARK_FOLDER).join(INDEX_PATH);
-        log::info!("Loading the index from file {}", index_path.display());
-        let file = File::open(&index_path)?;
-        let mut index = ResourceIndex {
-            id2path: HashMap::new(),
-            path2id: HashMap::new(),
-            collisions: HashMap::new(),
-            root: root_path.clone(),
-        };
+        let index_path = root_path
+            .as_ref()
+            .join(ARK_FOLDER)
+            .join(INDEX_PATH);
+        log::info!("Loading the index from file: {}", index_path.display());
 
-        // We should not return early in case of missing files
-        let lines = BufReader::new(file).lines();
-        for line in lines {
-            let line = line?;
-
-            let mut parts = line.split(' ');
-
-            let modified = {
-                let str = parts.next().ok_or(ArklibError::Parse)?;
-                UNIX_EPOCH.add(Duration::from_millis(
-                    str.parse().map_err(|_| ArklibError::Parse)?,
-                ))
-            };
-
-            let id = {
-                let str = parts.next().ok_or(ArklibError::Parse)?;
-                ResourceId::from_str(str)?
-            };
-
-            let path: String =
-                itertools::Itertools::intersperse(parts, " ").collect();
-            let path: PathBuf = root_path.join(Path::new(&path));
-            match CanonicalPathBuf::canonicalize(&path) {
-                Ok(path) => {
-                    log::trace!("[load] {} -> {}", id, path.display());
-                    index.insert_entry(path, IndexEntry { id, modified });
-                }
-                Err(_) => {
-                    log::warn!("File {} not found", path.display());
-                    continue;
-                }
-            }
-        }
+        let file = File::open(index_path)?;
+        let reader = BufReader::new(file);
+        let index: ResourceIndex = serde_json::from_reader(reader)?;
```

Comment:
Am I right that this changes the current file format to JSON?

In the `main` branch, index is stored in plain-text format:
```
1675087824710 58922-3276384608 Info/crdt_vs_ot.jpg
1675087824610 125308-3041567246 Info/Watson-Crick & Hoogsteen.png
1675087831180 386929-63288935 Screenshots/Screenshot_20210717_120856.jpg
1675087831440 945565-3484272718 Screenshots/Screenshot_20220111_040205_com.adobe.reader.jpg
1675087832200 57186-2804307164 Screenshots/lenovo-fedora-20210427-023231-183683953.png
1675087832990 107862-4031618865 Screenshots/lenovo-fedora-20210328-050520-182040364.png
1675087833830 343648-561346264 Screenshots/Screenshot_20210804_003308.jpg
1675087834130 257481-490338418 Info/Technologies.jpg
1675087846130 415620-3764250547 Gif/сitrus.gif
```

Reply:
I have found that storing and loading the same index with the `main` branch doesn't produce the same object. Not a matter of sorting but timestamps might be different as well. The code in the `main` branch for storing and loading contains more logic that just writing the struct to a file.
The idea here was to make `ResourceIndex` implement `Serialize`. with this setup, [we can serialize to various formats](https://serde.rs/#data-formats)

but It's up to you, we can revert to the old file format if you prefer it

Reply:
I like your idea, I just don't want to write huge JSON to disk.

Some compact format would be great: postcard, CBOR, what else could serve this purpose?

Reply:
On other hand, if we use automatic serialization, then the format will be excessive most likely: 2 mappings means creating two separate structures. Right now, we write only one table and derive 2 mappings from it during reading.

Reply:
> On other hand, if we use automatic serialization, then the format will be excessive 

Good point. I noticed that as well.

I now reverted `ResourceIndex::store()` and `ResourceIndex::load()` to the old code. I just kept the code to make them serializable to have this possibility later. 

I just wanted to add some notes about the current implementation of the 2 methods:
- we currently don't store or load information about collisions in `ResourceIndex`
- I noticed some inconsistencies when trying to store an object of `ResourceIndex` and then load it. I noticed some differences not only in order but also in `modified_time`. I think this may be a problem with how we parse the time

I could move this conversation to an issue and work on a separate PR to solve it

Reply:
@tareknaser this is interesting
> I noticed some inconsistencies when trying to store an object of ResourceIndex and then load it. I noticed some differences not only in order but also in modified_time. I think this may be a problem with how we parse the time

Could you tell more about the timestamps discrepancies? Probably steps to reproduce with `ark-cli`? Yeah, we could create separate issue if this is a bug indeed.

> we currently don't store or load information about collisions in ResourceIndex

Yes, but is it an issue? Collisions should be derivable from the table itself, right?

Reply:
Btw test case ` fn index_entry_order()` could help debug lines ordering.

Reply:
> Could you tell more about the timestamps discrepancies? Probably steps to reproduce with ark-cli? Yeah, we could create separate issue if this is a bug indeed.

I actually noticed this difference when I wrote a simple test
Here is the test to reproduce 
```rust
#[test]
fn resource_index_load_store() {
    let temp_dir = TempDir::new("arklib_test")
        .expect("Failed to create temporary directory");
    let temp_dir = temp_dir.into_path();

    create_file_at(
        temp_dir.to_owned(),
        Some(FILE_SIZE_1),
        Some(FILE_NAME_1),
    );
    let index = ResourceIndex::build(temp_dir.to_owned());


    index
        .store()
        .expect("Should store index successfully");


    let loaded_index = ResourceIndex::load(temp_dir.to_owned())
        .expect("Should load index successfully");

    // Assert that the loaded index is equal to the original index
    assert_eq!(index, loaded_index);
}
```

When I run this test, I get this output
```bash
thread 'index::tests::resource_index_load_store' panicked at src/index.rs:824:9:
assertion `left == right` failed
  left: ResourceIndex { id2path: {ResourceId { data_size: 10, crc32: 3817498742 }: "/private/var/folders/fg/kw558kp56h574pchd9zylc500000gn/T/arklib_test.VcRhfirveNvX/test1.txt"}, path2id: {"/private/var/folders/fg/kw558kp56h574pchd9zylc500000gn/T/arklib_test.VcRhfirveNvX/test1.txt": IndexEntry { modified: SystemTime { tv_sec: 1709340915, tv_nsec: 542881544 }, id: ResourceId { data_size: 10, crc32: 3817498742 } }}, collisions: {}, root: "/private/var/folders/fg/kw558kp56h574pchd9zylc500000gn/T/arklib_test.VcRhfirveNvX" }
 right: ResourceIndex { id2path: {ResourceId { data_size: 10, crc32: 3817498742 }: "/private/var/folders/fg/kw558kp56h574pchd9zylc500000gn/T/arklib_test.VcRhfirveNvX/test1.txt"}, path2id: {"/private/var/folders/fg/kw558kp56h574pchd9zylc500000gn/T/arklib_test.VcRhfirveNvX/test1.txt": IndexEntry { modified: SystemTime { tv_sec: 1709340915, tv_nsec: 542000000 }, id: ResourceId { data_size: 10, crc32: 3817498742 } }}, collisions: {}, root: "/private/var/folders/fg/kw558kp56h574pchd9zylc500000gn/T/arklib_test.VcRhfirveNvX" }
```

Notice how `tv_nsec` is different

Reply:
Looks like rounding.. If you could investigate why is it rounded, it would be great.

Reply:
I think I understand now
We're currently storing and loading timestamps in milliseconds, but `SystemTime` relies on `Timespec` under the hood which is:
```rust
struct Timespec {
    tv_sec: i64,
    tv_nsec: Nanoseconds,
}
```
I confirmed locally that if we store and parse the timestamp as nanoseconds, we don't run into this issue. I'll double-check before pushing the code.